src/main/java/org/springframework/integration/aws/outbound/KplMessageHandler.java

```diff
@@ -368,11 +429,50 @@ private void setGlueSchemaIntoUserRecordIfAny(UserRecord userRecord, Message<?>
 	}
 
 	private CompletableFuture<UserRecordResponse> handleUserRecord(UserRecord userRecord) {
+		if (this.maxInFlightRecords != -1) {
+			waitForCapacityInUserRecordsBuffer();
+		}
+
 		ListenableFuture<UserRecordResult> recordResult = this.kinesisProducer.addUserRecord(userRecord);
 		return listenableFutureToCompletableFuture(recordResult)
 				.thenApply(UserRecordResponse::new);
 	}
 
+	private void waitForCapacityInUserRecordsBuffer() {
+		var kplOutstandingRecordsCount = this.kinesisProducer.getOutstandingRecordsCount();
+		var attempts = 1;
+		var sleepDuration = this.maxInFlightRecordsBackoffDuration;
```

Comment:
Doesn't look like this variable is used somehow in the logic.

Reply:
Its a great catch. Thank you @artembilan.